### DISCUSSION TITLE
[bug report] Pull request comments API endpoint is empty

### DISCUSSION THREAD

**Discussion by tryvin** [community/community#42725] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

Hey Guys, so, I think that after this last API update the pull request comments API is broken. I noticed this because our "Danger-JS" settings is set to remove previous comments it created, and it has been a while since it hasn't deleted them.

When I investigated this further, using a personal token from my own account, I noticed that all `repos/ORG HERE/REPO HERE/pulls/PR NUMBER HERE/comments` were returning an empty JSON, this is one of the tests github request id: `x-github-request-id: 3787:35B4:8BA2C0:9FABF2:63AB544B`

Can someone check this out?

**Comment by tryvin** [comment]
Hi @queenofcorgis @rileybroughten sorry to bother you guys, but this is kinda "High Priority" as the API is not working as it should

**Reply by Vikramjayanthi** [reply]
Same problem here!

**Comment by faoiseamh** [comment]
I'm seeing the same thing as well. This appears to be a major + blocking regression of a primary API function. @queenofcorgis please help us get some visibility here

**Comment by queenofcorgis** [comment]
Hi there! I passed this along to the Pull Requests team when first reported, I will follow up with them now that there has been an additional report - if you have any more details that may be relevant, please share them here.

**Comment by tryvin** [comment]
@queenofcorgis Any news?

**Reply by queenofcorgis** [reply]
I've connected with some of the engineers - are you having trouble accessing comments for all PRs?

Try this to see if it works: 

```
gh api \
  -H "Accept: application/vnd.github+json" \
  repos/rails/rails/pulls/46972/comments
```

**Reply by queenofcorgis** [reply]
```
curl \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer <PERSONAL ACCESS TOKEN HERE>"\
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/rails/rails/pulls/46972/comments
```
(You'll need to generate a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) though) (edited)

**Comment by tryvin** [comment]
@queenofcorgis yes, all the PRs are having this empty comments issue, with my token, with the robot token, it doesn't matter.

**Comment by gpressutto5** [comment]
@queenofcorgis For me, it returns 6 comments, but it is not what we expect this endpoint to return. It is returning only the review comments. I expected to also see these comments:

> [This is awesome!](https://github.com/rails/rails/pull/46972#issuecomment-1379910630)

> [I will have to make the same changes to the API docs and remove that new-line](https://github.com/rails/rails/pull/46972#issuecomment-1379910630)

But I only get the comments from this review:

<img width="534" alt="image" src="https://user-images.githubusercontent.com/12385501/212486459-ffe6f3ae-bc11-4e30-8e0f-9ddef87995c1.png">

**Comment by gpressutto5** [comment]
@queenofcorgis Any news?

**Comment by tryvin** [comment]
@queenofcorgis any news?

**Comment by simon-enplore** [comment]
@queenofcorgis Any news?

**Comment by simon-enplore** [comment]
@tryvin @gpressutto5 @faoiseamh
Have any of you found any reasonable bypass for the issue? Using GraphQl/ other endpoint or something?

**Comment by queenofcorgis** [comment]
@tryvin @simon-enplore I'm continuing to check in with the Pull Requests team, but no updates as of yet - really appreciate the patience!

**Comment by queenofcorgis** [comment]
Reporting back, we tagged in a new team member who reviewed this thread and there may be some confusion here. 

Issue comments and commit comments **_will not_** be returned. Only Pull Request review comments will be returned with the API. [Check out our docs](https://docs.github.com/en/rest/pulls/comments?apiVersion=2022-11-28#about-pull-request-review-comments) for more details. 

That being said, if you still continue to experience an issue with review comments, please open a ticket with our [Support team](https://support.github.com/).

**Reply by tryvin** [reply]
@queenofcorgis So I think something changed at the point, as this endpoint was being used by Danger to grab all the PR comments.

So, let's clarify things here:

- All the comments made by users into a PR will be returned via the `issue comments` endpoint
- Pull request reviews comments (the ones inside a single pull request review) will be returned via the `Pull request reviews comments` endpoint

Is that right?

**Reply by queenofcorgis** [reply]
* All Issue comments will be returned by the Issue comment endpoint
* Your second bullet is correct 😄

**Comment by chvolk** [comment]
The pulls endpoint is still returning empty data for me. Followed the docs and using curl:

 curl -L -H "Accept: application/vnd.github+json" -H "Authorization: Bearer <MY AUTH KEY>" -H "X-GitHub-Api-Version: 2022-11-28" https://api.github.com/repos/chvolk/Fullmetal/pulls.

Is there an update on this?

**Comment by jerry-b** [comment]
What I am experiencing is that the issues/.../comments endpoint **only returns comments that were made by the PR author.**  Comments by others can be seen in the PR but are not returned.

I can reliably reproduce this.

That may be why some are saying that it is empty - since you will see empty if the author did not leave a PR level (issue) comment.
@queenofcorgis

**Reply by queenofcorgis** [reply]
@jerry-b would you mind sharing the request you are making so we could take a look please?

**Reply by jerry-b** [reply]
https://api.github.com/repos/HomeVision/price/issues/9095/comments
 https://api.github.com/repos/HomeVision/price/pulls/9095/comments
 
For example, in this specific PR there are 3 review (pulls) comments and two issue comments. The APIs return all of those comments except the issue comment made by someone besides the author. 

Thanks @queenofcorgis !

**Reply by jerry-b** [reply]
I also posted a new comment as I realized this one was old and possibly not the same problem.
https://github.com/orgs/community/discussions/114441

**Reply by JanKoszewski** [reply]
Hi @jerry-b, I'm an engineer on the pull requests team. I think that for the second issue for which you've created the [new discussion](https://github.com/orgs/community/discussions/114441), the [issue comment API](https://docs.github.com/en/rest/issues/comments?apiVersion=2022-11-28) is being used in a manner for which it isn't built. I also think that there might be a fundamental misunderstanding about the type of comments that are being created on a pull request and their availability between the pull request review comment api and the issue comment api. For better or worse I don't have access to the `HomeVision/price` repo (and therefore the api queries you've listed above), but I'll try using this `rails/rails` pull request as an example: https://github.com/rails/rails/pull/51339

On this pull request we can see a few 'comments', but they're actually distinctly different types of comments on the backend. Pull request review comments are those that are made on the file or diff of a pull request itself; the body text of pull request reviews also appear like they're comments / alongside comments on the frontend, but they're created as a part of the process of actually issuing a review on a pull request and are not treated as pull request review comments. Here's more specifically what I mean: 

<img width="879" alt="Screenshot 2024-03-27 at 2 49 38 PM" src="https://github.com/community/community/assets/811977/14a85aa7-5e80-49c1-b44f-e99745924866">

If I make a request to the pull request comment REST api for this pull request, (`https://api.github.com/repos/rails/rails/pulls/51339/comments` specifically) these pull request comments (and **only** other pull request comments) appear in the response:

```
➜  ~ curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer ..." \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/rails/rails/pulls/51339/comments
 [
  {
    "url": "https://api.github.com/repos/rails/rails/pulls/comments/1528675373",
    "pull_request_review_id": 1943253181,
    "id": 1528675373,
    ...
    "body": "Please remove this file, configs should be kept local.",
    "created_at": "2024-03-18T14:28:41Z",
    "updated_at": "2024-03-18T14:40:00Z",
    ...
  },
  ...
  {
    "url": "https://api.github.com/repos/rails/rails/pulls/comments/1528697601",
    "pull_request_review_id": 1943290842,
    "id": 1528697601,
    ...
    "body": "Gah, sorry!",
    "created_at": "2024-03-18T14:41:02Z",
    "updated_at": "2024-03-18T14:41:02Z",
    ...
  },
  ...
]
```

Body text of reviews do not show up in either the pull request review comment API or the issue comment api. Moreover, since these are distinctly pull request review comments, they do not show up at all when I make a request to the issue comment REST api (`https://api.github.com/repos/rails/rails/issues/51339/comments` specifically). What _does_ show up in the issue comment api are comments made literally on the 'issue' portion of the pull request and _not_ on a specific file or part of the pull request diff - namely comments like these:

<img width="923" alt="Screenshot 2024-03-27 at 2 53 18 PM" src="https://github.com/community/community/assets/811977/d918146e-6563-4a58-b733-b344d2ae91e9">

Which, again, we see when we make a request to the issue comment REST api:

```
➜  ~ curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer ..." \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/rails/rails/issues/51339/comments
[
  ...
  {
    "url": "https://api.github.com/repos/rails/rails/issues/comments/2005752431",
    "html_url": "https://github.com/rails/rails/pull/51339#issuecomment-2005752431",
    "issue_url": "https://api.github.com/repos/rails/rails/issues/51339",
    "id": 2005752431,
    ...
    "body": "> I added .vscode to .gitignore; please let me know if that's acceptable\r\n\r\nPer the comment at the top of that file, it is not.",
    ...
  },
  {
    "url": "https://api.github.com/repos/rails/rails/issues/comments/2007785788",
    "html_url": "https://github.com/rails/rails/pull/51339#issuecomment-2007785788",
    "issue_url": "https://api.github.com/repos/rails/rails/issues/51339",
    "id": 2007785788,
    ...
    "body": "Made all the requested changes. I think the current CI failure is something flakey:\r\n\r\n```\r\nUsing trilogy\r\nrake aborted!\r\nActiveRecord::ConnectionNotEstablished: Connection refused - trilogy_connect - unable to connect to mysql:3306 (ActiveRecord::ConnectionNotEstablished)\r\n```\r\n\r\nI also can't see a way to restart the build, I guess because of admin permissions. ",
    ...
  }
]
```

Does that explanation help? I've been unable to corroborate your finding that either the pull request review comment or issue comment APIs return comments only made by the author, and since I again can't see or make requests to the `HomeVision/price` repo, I can't independently see what the issue you're encountering might be. Is it possible for you to recreate the error you're seeing in a public repository?

**Reply by mridulghanshala** [reply]
@JanKoszewski thanks for that is there a way to actually access the `body` of a pr then in ur example: `Thanks, I had a few suggestions below` ?

**Reply by Ahmed-Habeb26102004** [reply]
If I want a template to use, where can I find it?

**Reply by gerazenobi** [reply]
@JanKoszewski thanks for the detailed answer and clarifications. 

> @JanKoszewski thanks for that is there a way to actually access the `body` of a pr then in ur example: `Thanks, I had a few suggestions below` ?

I was also interested in this: how do we get that information `body of a review` since it isn't an `issue comment` nor a `review comment` ?

**Reply by joshuataylor** [reply]
Yeah, seems like this is still broken, I am trying to find a way to get all comments, reviews, etc for a PR and this seems to be missing in the reviews endpoint.

The GitHub API seems a maze of GraphQL, REST, where some things you can do through one endpoint, but not through another.
